### ISSUE TITLE
SA1402 does not take topLevelType setting into account for records

### ISSUE THREAD

**Issue by appel1** [DotNetAnalyzers/StyleCopAnalyzers#3672]
In our stylecop json we have:
```
"maintainabilityRules": {
   "topLevelTypes": [ "class" ]
}
````

But still it warns about files containing more than one record. It's fine if the default for topLevelTypes has changed to include records  but not being able to configure it at all is inconvenient.

**Comment by bjornhellander** [comment]
The behaviour you are experiencing is what was decided in #3542. Record, record class and record struct were simply ignore before that, if I am not mistaken.

Can you describe why you think it would be good to be able to configure the behaviour for class and record declarations separately?

**Comment by appel1** [comment]
We primarily use records (and rarely record struct) for small dtos with very little logic if any. So we treat them as other small things like delegates, small interfaces, small structs and the like and put the together with the class that produces or consumes them or many of them in a separate file if that makes more sense. It just clutters things up to put these one liners into separate files. 

If record and record struct were separate and the default for topLevelTypes was changed to [ "class", "record" ] instead you could get both the current behavior and allow us to configure it in the way we prefer. Right now our only sensible option is to turn off SA1402, or suppress it all over the place.

**Comment by appel1** [comment]
It doesn't really make sense to allow structs in the same file as other declaration or a class but not record.

Struct:
```csharp
    public readonly struct Something : IEquatable<Something>
    {
        public Something(int a, int b, int c)
        {
            A = a;
            B = b;
            C = c;
        }

        public int A { get; }

        public int B { get; }

        public int C { get; }

        public override bool Equals(object? obj)
        {
            return obj is Something something && Equals(something);
        }

        public bool Equals(Something other)
        {
            return A == other.A &&
                   B == other.B &&
                   C == other.C;
        }

        public override int GetHashCode()
        {
            int hashCode = 793064651;
            hashCode = hashCode * -1521134295 + A.GetHashCode();
            hashCode = hashCode * -1521134295 + B.GetHashCode();
            hashCode = hashCode * -1521134295 + C.GetHashCode();
            return hashCode;
        }

        public static bool operator ==(Something left, Something right)
        {
            return left.Equals(right);
        }

        public static bool operator !=(Something left, Something right)
        {
            return !(left == right);
        }
    }
```

Record:
```csharp
public readonly record Something(int A, int B, int C);
```

**Comment by sharwell** [comment]
For now, I'm comfortable with the current approach. Occasionally there are short records like the one shown above, but in practice they often end up with one or more members added which make them more like small types.

**Comment by appel1** [comment]
By that logic struct should also be handled as class since they always have members. The default should also be changed to include both class and struct. 

I'm fine with that but would like a setting to exclude struct, record and record struct. 

If it is impossible to add new fields to topLevelTypes perhaps a new excludeFromTopLevelTypes can be added that is empty by default and where you can add struct, record and recordstruct.

**Comment by sharwell** [comment]
> The default should also be changed to include both class and struct.

I would have absolutely agreed with this if we created this rule. However, it was grandfathered in from StyleCop Classic so we kept the default behavior from prior work to ease migration.

**Comment by appel1** [comment]
A way to configure this would be nice either way. 

Until then we'll have to turn it off because fixing the violations would lead to an unwieldy amount of one liner files.

**Comment by appel1** [comment]
A way to configure this would be nice either way. 

Until then we'll have to turn it off because fixing the violations would lead to an unwieldy amount of one liner files.

**Comment by sharwell** [comment]
You could put a `#pragma warning disable SA1402` at the top of the (presumably) small number of files with multiple declarations.

**Comment by appel1** [comment]
@sharwell We opted to disable SA1402 and create our own that only requires class and struct to be in separate files instead of hundreds of `#pragmas`. If you change your mind in the future we can just revert that change.

**Comment by lennartb-** [comment]
Yeah, more configuration options would have been great. We have many command handler classes which include the commands as records with 1- 3 members (i.e. one-liners), adding a suppression to each file is not nice. I guess we'll disable SA1402 for these particular folders in .editorconfig.

**Comment by danadesrosiers** [comment]
I think the change made to SA1402 in #3542 was sensible, but it should only apply to records with a body.  Positional records with no body, like the example given by @appel1 above, should not be subject to this rule.  In many cases, it's silly to put one line of code in a separate file.
